err on invalid feature flag

stefan0xC opened the pull request:
As mentioned in the PR https://github.com/dani-garcia/vaultwarden/pull/4168#discussion_r1431892943 we have to `err!` to disallow unknown flags otherwise all flags are allowed.

I've just tested it and the warning is not even displayed. Might be the same reason as
https://github.com/dani-garcia/vaultwarden/blob/890e668071cffe2833834348e19bbef3c061d014/src/config.rs#L128

BlackDex commented:
The problem with that is, if we remove a flag, we will break those setups.

stefan0xC commented:
Let me quote myself:

> In order not to break too many installations we probably should not remove features from `KNOWN_FLAGS` that we have enabled by default (like `fido2-vault-credentials`) or only if there's an important reason to do so or we really want to remove it. (I think the other flags, which we don't enable by default, would be fair game as users would be responsible if they change the defaults).

We could also inform our users in the release notes about removed flags if we do remove them.

stefan0xC commented:
Let me put it in other words: Without this change there is no removing flags and we could also geht rid of the validation.

BlackDex commented:
I still do not like using err! here. My main point is.
If a user auto updates Vaultwarden, and a flag has been removed from the known flags list, it will break that setup and the instance is not reachable anymore. So, for me, that is a no-go.

What is happening currently is also not ok i think.
It should filter out the unknown flags and return only a list of valid items and warn/println about the unknown ones.

Currently the latest Self-Hosted is returning:
```json
"trusted-device-encryption": true,
"fido2-vault-credentials": true
```

So the actions i think we need to take to fix this are:
1. Still warn/println about unknown flags
2. Filter those unknown flags, and only return valid flags
3. Maybe have a default list like `fido2-vault-credentials`, and always add that, unless it is excluded with `^fido2-vault-credentials` for example.
  That way we are still able to add new flags by default, and allow users to override them if they do not want it.

stefan0xC commented:
Why do we have to remove something from the `KNOWN_FLAGS`?

BlackDex commented:
Because, if those flags are not needed anymore, we probably also do not want them to be added.
It might even cause troubles in the future or something if they are still in there.
What if they re-use a flag for some reason. They know when they removed a flag and were not using it for a while, and can just re-use a previous flag for any other reason.

stefan0xC commented:
I'd say in those cases it would be fine to break a user's invalid configuration, is it not? :raised_eyebrow: 

I don't see the point in making sure it doesn't break (by filtering out unknown flags) and somehow still warn a user (which doesn't currently work). I mean, why would anyone fix their invalid config, if it just keeps working?

BlackDex commented:
The only way i would see `err!` work is if we do not remove known features, and when we do, we need to filter those out, and have at least those a warn, and no err.

Also, best to also parse all flags, and create an err with all invalid flags instead of break on the first invalid one.

stefan0xC commented:
Okay, I've rewritten the check so it prints all invalid flags and also tells users what flags are supported:

```
Error loading config:
  Unrecognized experimental client feature flags: ["passwordless-login", "bla"].

Please ensure all feature flags are spelled correctly and that they are supported in this version.
Supported flags: ["autofill-overlay", "autofill-v2", "browser-fileless-import", "fido2-vault-credentials"]
```

Regarding the other change, you've suggested: I've added a comment so that we don't forget about dealing with removed flags.
